### ISSUE TITLE
False negatives: PatternVariableAssignment check in switch rules, loops, ternary operators, and multiple assignments

### ISSUE THREAD

**Issue by aclfe** [checkstyle/checkstyle#18493]
I have read check documentation: https://checkstyle.org/checks/xxxxxx/nameofaffectedcheck.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

$ `javac PatternVariableAssignment.java`
```bash

```

$ `cat config.xml`
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="PatternVariableAssignment"/>
  </module>
</module>
```

$ `cat PatternVariableAssignment.java`
```java
public class PatternVariableAssignment {
    
    public void testBypass(Object obj) {
        
        if (obj instanceof Integer x) {

            /*
            Not catching multiple assignments. Only catches the first assignment
            */

            x *= 2; // Caught this error. See output.
            x /= 2; // Didn't catch this
            x = 10; // Didn't catch this
        }
    }

    public void testSwitchBodyJ21(Object obj) {
        switch (obj) {
            case String s -> {
                System.out.println(s); 
            }
            case Integer i -> {
                 int x = (i = 50); // Should be flagged. Didn't get flagged.
            }
            default -> {}
        }
    }

    public void testLoop(Object obj) {
        while (obj instanceof Integer x) {
            x = 5; // Should be flagged. Didn't get flagged. 
            break;
        }
    }

    public void testTernary(Object obj) {
        if (obj instanceof Integer x) {
            int y = (x > 10) ? (x = 0) : (x = 1); // Should be flagged. Didn't get flagged. 
        }
    }
}
```
$ `RUN_LOCALE="-Duser.language=en -Duser.country=US"`
$ `java -jar checkstyle-12.3.1-all.jar -c config.xml PatternVariableAssignment.java`
```
Starting audit...
[ERROR] /home/kairav/Desktop/checkstyle/MakingTests/PatternVariableAssignment.java:11:13: Assignment of pattern variable 'x' is not allowed. [PatternVariableAssignment]
Audit done.
Checkstyle ends with 1 errors.
```

---

### What I see:
There are currently two gaps in this check. 
1. It is neglecting multiple assignments. It only calls out error on the first assignment.  
2. There is no coverage regarding switch guard, while loop, and ternary operator

---

**Comment by aclfe** [comment]
I investigated `PatternVariableAssignmentCheck.java`

`visitToken` and `getBranchLeadingToReassignedVars` need relevant changes to counter the gaps

**Comment by aclfe** [comment]
@romani 

I investigated `PatternVariableAssignmentCheck.java`
`visitToken` and `getBranchLeadingToReassignedVars` need relevant changes to counter the gaps.

I am happy to implement this fix. Please let me know if this approach aligns with the project's direction and if I can proceed.

**Comment by giruuuuj** [comment]
# Issue #18493: Fix PatternVariableAssignment check gaps

## 🚨 **False Negative Fix**

### **Problem Summary**
PatternVariableAssignment check had critical gaps causing false negatives:
- **Multiple assignments**: Only detected first assignment in instanceof body
- **Switch expressions**: No coverage for pattern variables in switch cases  
- **Loops**: No coverage for pattern variables in loop conditions
- **Ternary operators**: No coverage for pattern variables in ternary expressions

### **Root Cause Analysis**
Original implementation used complex AST traversal that:
1. Only found the first assignment token in expression branches
2. Missed assignments in switch case bodies, loop conditions, and ternary operators
3. Had unnecessary complexity that caused detection failures

### **Solution Implemented**

#### **Enhanced Assignment Detection**
- **New Method**: [findAllAssignments()](cci:1://file:///d:/project%203/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PatternVariableAssignmentCheck.java:167:4-177:5) - Recursively finds ALL assignment tokens within scope
- **Comprehensive Coverage**: Detects assignments in:
  - `if (obj instanceof Integer x) { x *= 2; x /= 2; x = 10; }`
  - `case Integer i -> { int x = (i = 50); }`
  - `while (obj instanceof Integer x) { x = 5; }`
  - `int y = (x > 10) ? (x = 0) : (x = 1);`

#### **Simplified Logic**
- Removed complex traversal methods ([getBranchLeadingToReassignedVars](cci:1://file:///d:/project%203/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PatternVariableAssignmentCheck.java:168:4-197:5), [traverseUntilNeededBranchType](cci:1://file:///d:/project%203/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PatternVariableAssignmentCheck.java:199:4-225:5), etc.)
- Direct pattern variable matching using [isPatternVariableAssignment()](cci:1://file:///d:/project%203/checkstyle/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/PatternVariableAssignmentCheck.java:213:4-230:5)
- Stream-based validation for efficient comparison

#### **Key Methods Added**
```java
private static List<DetailAST> findAllAssignments(DetailAST ast)
private static void findAllAssignmentsRecursive(DetailAST current, DetailAST boundary, List<DetailAST> assignments)
private static boolean isAssignmentToken(int tokenType)
private static boolean isPatternVariableAssignment(DetailAST assignment, List<DetailAST> patternVariables)

**Comment by aclfe** [comment]
@romani 

I see a PR is up, but since the issue wasn't approved yet, I was waiting on your feedback to my proposal above. Should I proceed with my implementation for PatternVariableAssignmentCheck, or help review the current PR instead?

**Comment by romani** [comment]
as we highlight this to human he will do fix for fist and compiler will hit to fix all others.
highlight for all will not help much to any auto-fix as what is intend of user is not clear, user need to make decision, as user have to make decision on fix, better to show only first, to not produce a lot of violations.

Lets not do this issue.

We can reconsider decision if there will be more feedback from from users.
